packages/flutter/lib/src/material/text_selection_toolbar.dart

```diff
@@ -0,0 +1,721 @@
+      overflowOpen: overflowOpen,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _TextSelectionToolbarItemsLayoutRenderBox renderObject) {
+    renderObject
+      ..isAbove = isAbove
+      ..overflowOpen = overflowOpen;
+  }
+
+  @override
+  _TextSelectionToolbarItemsLayoutElement createElement() => _TextSelectionToolbarItemsLayoutElement(this);
+}
+
+class _TextSelectionToolbarItemsLayoutElement extends MultiChildRenderObjectElement {
+  _TextSelectionToolbarItemsLayoutElement(
+    MultiChildRenderObjectWidget widget,
+  ) : super(widget);
+
+  static bool _shouldPaint(Element child) {
+    return (child.renderObject!.parentData! as ToolbarItemsParentData).shouldPaint;
+  }
+
+  @override
+  void debugVisitOnstageChildren(ElementVisitor visitor) {
+    children.where(_shouldPaint).forEach(visitor);
+  }
+}
+
+class _TextSelectionToolbarItemsLayoutRenderBox extends RenderBox with ContainerRenderObjectMixin<RenderBox, ToolbarItemsParentData> {
+  _TextSelectionToolbarItemsLayoutRenderBox({
+    required bool isAbove,
+    required bool overflowOpen,
+  }) : assert(overflowOpen != null),
+       assert(isAbove != null),
+       _isAbove = isAbove,
+       _overflowOpen = overflowOpen,
+       super();
+
+  // The index of the last item that doesn't overflow.
+  int _lastIndexThatFits = -1;
+
+  bool _isAbove;
+  bool get isAbove => _isAbove;
+  set isAbove(bool value) {
+    if (value == isAbove) {
+      return;
+    }
+    _isAbove = value;
+    markNeedsLayout();
+  }
+
+  bool _overflowOpen;
+  bool get overflowOpen => _overflowOpen;
+  set overflowOpen(bool value) {
+    if (value == overflowOpen) {
+      return;
+    }
+    _overflowOpen = value;
+    markNeedsLayout();
+  }
+
+  // Layout the necessary children, and figure out where the children first
+  // overflow, if at all.
+  void _layoutChildren() {
+    // When overflow is not open, the toolbar is always a specific height.
+    final BoxConstraints sizedConstraints = _overflowOpen
+      ? constraints
+      : BoxConstraints.loose(Size(
+          constraints.maxWidth,
+          _kToolbarHeight,
+        ));
+
+    int i = -1;
+    double width = 0.0;
+    visitChildren((RenderObject renderObjectChild) {
+      i++;
+
+      // No need to layout children inside the overflow menu when it's closed.
+      // The opposite is not true. It is necessary to layout the children that
+      // don't overflow when the overflow menu is open in order to calculate
+      // _lastIndexThatFits.
+      if (_lastIndexThatFits != -1 && !overflowOpen) {
+        return;
+      }
+
+      final RenderBox child = renderObjectChild as RenderBox;
+      child.layout(sizedConstraints.loosen(), parentUsesSize: true);
+      width += child.size.width;
+
+      if (width > sizedConstraints.maxWidth && _lastIndexThatFits == -1) {
+        _lastIndexThatFits = i - 1;
+      }
+    });
+
+    // If the last child overflows, but only because of the width of the
```

Comment:
### Last child width
No, the relative width shouldn't matter.  This is about the case where the overflow button causes the overflow, like:

|--------- width ---------|
[button 1] [button 2] [overflow button]

If we just get rid of the overflow button, then everything fits and we don't need an overflow button.

### Wrong Constraints
In response to another comment above (https://github.com/flutter/flutter/pull/69428#discussion_r538878026), I noticed that I should be using padding to set the constraints of the menu properly.  Does that cover your concern here?

### Consistent Size
Good catch, I've created an issue for this: https://github.com/flutter/flutter/issues/72023.  I spent a bunch of time trying to fix this right now, but I don't think I can align the text in a TextButton.  This problem existed before this PR, so I'll come back to it in that issue later.

Reply:
Sorry I wasn't clear. The overflow button is always the first child so in the case that it is extremely wide and takes up more horizontal space than button 1 and button 2 combined, we need to check more than the last child. 

i.e., 
children in layout order:
|--------- width ---------|
[wiiiiiiiiiiiiiiiiiiiiiiiiiide overflow button][button 1] [button 2]

Reply:
Ah that makes sense, you're right.  I'm creating a separate issue for this as well since it's not new to this PR and should be pretty unlikely: https://github.com/flutter/flutter/issues/72075